File: p003/p003.cpp
#include <iostream>
#include <vector>
#include <math.h>
#include <algorithm>
using namespace std;

void primes(vector<long> &px, long n) {
    px.push_back(2);
    for(long i=3; i<n; i+=2) {
        long limit = (long)sqrt(i);
        bool isPrime = true;
        for(long j=3; j<limit; j+=2)
        {
            if(i % j == 0)
            {
                isPrime = false;
                break;
            }
        }
        if(isPrime) {
            px.push_back(i);
        }
    }
}

void factories(vector<long> &fx, long n) {
    int k = n;
    vector<long>::iterator it;
    vector<long> px;
    primes(px, (long)sqrt(n));

    for(it=px.begin(); it!=px.end(); ++it) {
        long p = *it;
        if(n % p == 0) {
            k /= p;
            fx.push_back(p);
        }
        if(k == 0) {
            return;
        }
    }
}

int main() {
    vector<long> fx;
    vector<long>::iterator it;
    //factories(fx, 13195);
    //for(it=fx.begin(); it!=fx.end(); ++it) {
    //    cout << *it << " ";
    //}
    factories(fx, 600851475143);
    cout << *max_element(fx.begin(), fx.end()) << endl;
    return 0;
}

File: p003/p003.go
package main

import "fmt"
import "math"

func primes(n int) []int {
	px := make([]int, 1)
	px[0] = 2
	for i := 3; i <= n; i += 2 {
		limit := int(math.Sqrt(float64(i)))
		is_prime := true
		for j := 3; j <= limit; j += 2 {
			if i%j == 0 {
				is_prime = false
				break
			}
		}
		if is_prime {
			px = append(px, i)
		}
	}
	return px
}

func factories(n int) []int {
	fx := make([]int, 0)
	px := primes(int(math.Pow(float64(n), 0.5)))
	k := n
	for _, x := range px {
		if n%x == 0 {
			k /= x
			fx = append(fx, x)
		}
		if k == 0 {
			return fx
		}
	}
	return fx
}

func main() {
	//fmt.Printf("%v\n", factories(13195))
	fx := factories(600851475143)
	fmt.Printf("%d\n", fx[len(fx)-1])
}
